### ISSUE TITLE
ICS23: Add support for SHA-512/256 hash function

### ISSUE THREAD

**Issue by kostko** [cosmos/ibc#554]
The current Merkle proof abstraction defined by ICS 23 (or at least the major [Go](https://github.com/confio/ics23/blob/f173e6211efbf546bf9fcdc5cc29e5e055151666/go/ops.go#L112-L136) and [Rust](https://github.com/informalsystems/ibc-rs/blob/11f387959386e85d34bef1afddc1ee411b7b7214/proto/src/prost/ics23.rs#L249-L258) implementations) and used for state commitment proofs does not seem to include support for the SHA-512/256 hash function (actually it seems that the supported hash functions are not actually part of the [spec if I'm reading this right](https://github.com/cosmos/ibc/tree/master/spec/core/ics-023-vector-commitments), so not sure where this should go).

The Oasis Protocol uses a [custom merkelized binary patricia trie data structure](https://github.com/oasisprotocol/oasis-core/tree/master/go/storage/mkvs) instantiated with SHA-512/256 as the hash function so it would be great if we could see support for it in IBC. The change to support an additional hash function should be straightforward.

We may need other changes related to node encoding, but this is something that is orthogonal and needs further exploration.

**Comment by ethanfrey** [comment]
@kostko The supported hash functions are defined here: https://github.com/confio/ics23/blob/master/proofs.proto#L5-L13

The implementation to validate any proof in the ics23 format is also included in that repo. Note that the Cosmos IAVL tree uses SHA256 from this repo, so you are only discussing sha2-512 I believe.

You should be able to define a custom spec for your binary patricia trie. However, it requires a normal format. Ethereum's Patricia Trie which does arbitrary compression as well as variable encoding leafs as separate nodes are part of their parent based on length, proved impossible to include in the ics23 spec (it required tons of custom logic to verify). The Ethereum 2.0 SMT is much more regular and should be able to be proven by the current spec.

The first step is to find a document explaining exactly how one will validate a proof from Oasis's mkvs (I clicked the link above to the code and saw no docs).

**Comment by kostko** [comment]
> The supported hash functions are defined here: https://github.com/confio/ics23/blob/master/proofs.proto#L5-L13

Right, but the text of the specification (in this repository) does not specify any of the supported hash functions. Are those implementation-specific details? To me it seems like these should be standardized otherwise interoperability is harder.

> The implementation to validate any proof in the ics23 format is also included in that repo. Note that the Cosmos IAVL tree uses SHA256 from this repo, so you are only discussing sha2-512 I believe.

Yeah I know about IAVL and that it uses SHA256. We are using SHA-512/256 which is the truncated version of SHA-512 with its own IVs so you cannot just use SHA-512 (which is supported in the above implementation under enum tag `2`).

> You should be able to define a custom spec for your binary patricia trie. However, it requires a normal format.

Can you explain what you mean by a "normal format"? Is there some document explaining the Merkle proof abstraction you are using? Also this doesn't seem to be part of the specification of ICS 23, the specification actually is very high level and does not actually define anything except some abstract interface?

**Comment by ethanfrey** [comment]
> We are using SHA-512/256 which is the truncated version of SHA-512 with its own IVs so you cannot just use SHA-512 (which is supported in the above implementation under enum tag 2).

I have never seen this algorithm before. Is it only used in your chain or are there standard docs? Can you provide links to stable Go, Rust, and Typescript implementations for it? If this is a standard, supported algorithm, I am happy to accept a PR adding it to the confio/ics23 repo.

> Also this doesn't seem to be part of the specification of ICS 23, the specification actually is very high level and does not actually define anything except some abstract interface?

Yeah, I am not involved in that, but did write confio/ics23 quite some time ago. Just look at the specs we write there descriping iavl tree and tendermint merkle proofs. It is a "config format" to define the steps needed to validate.

**Comment by kostko** [comment]
> I have never seen this algorithm before. Is it only used in your chain or are there standard docs? Can you provide links to stable Go, Rust, and Typescript implementations for it? If this is a standard, supported algorithm, I am happy to accept a PR adding it to the confio/ics23 repo.

Yes, this is a standard algorithm defined in the same standard that defines the SHA-2 family of hash functions (see [FIPS 180-4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)). There are a number of implementations available, including in the [Go standard library](https://golang.org/pkg/crypto/sha512/).

**Comment by kostko** [comment]
@ethanfrey I've opened https://github.com/confio/ics23/pull/39 which adds SHA-512/256 hash op and implements the missing FIXED32_LITTLE length op.

**Comment by kostko** [comment]
Also can someone point to any Rust implementations of ICS-23? The `ibc-rs` repository only seems to have the types but I don't see any verification logic?

**Comment by ethanfrey** [comment]
The rust implementation is inside confio/ics23. As is the Go and TypeScript. They are all kept in the same repo to ensure they stay in sync.

**Comment by kostko** [comment]
The PR should have all the implementations now.
